### ISSUE TITLE
Why the license GPL-3.0 ?

### ISSUE THREAD

**Issue by stephanoparaskeva** [typesense/typesense#116]
GPL is widely considered a bad license for Open Source. Competitors of this project seem to be using the MIT license which is great. Why is this project using a license like this, and what does that mean for the future?

A company may typically do this is they want to charge for a better license etc. Similar to what MongoDB has done in the past.

**Comment by kishorenc** [comment]
@stephanoparaskeva Thanks for reaching out. Could you please expand on what you’re looking to do and how GPL 3 is preventing you from doing that? All our client libraries which will be directly integrated with your code is Apache 2 licensed. Definitely want to address your concerns here.

**Comment by stephanoparaskeva** [comment]
The problem would be general I think. GPL infers that copy-left can be applied to any derivative work and is ambiguous as to what 'derivative' means. For that reason I think many companies avoid using software that share this license as well as licenses like the AGPL. Is it possible to ask what the reasoning is to use this license rather than anything else?

@kishorenc

**Comment by kishorenc** [comment]
From our experience companies are generally concerned when libraries they use are GPL licensed, since library code is directly integrated into their code and will lead to derivative work and trigger GPL compliance. However, Typesense Server is server software and we expect users to typically run it as a separate daemon, and not integrate it with their own code. GPL covers and allows for this use case generously (eg: Linux is GPL licensed). Now, AGPL is what makes server software accessed over a network result in derivative work and not GPL. And for that reason we’ve opted to not use AGPL for Typesense.

Now, if someone makes modifications to Typesense server, GPL actually allows you to still keep the modifications to yourself as long as you don't distribute the modified code. So a company can for example modify Typesense server and run the modified code internally and still not have to open source their modifications, as long as they make the modified code available to everyone who has access to the modified software (their employees).

Now, if someone makes modifications to Typesense server and distributes the modifications, that's where GPL kicks in. Given that we’ve published our work to the community, we’d like for others’ modifications to also be made open to the community in the spirit of open source. We use GPL for this purpose. Other licenses would allow our open source work to be modified, made closed source and distributed, which we want to avoid with Typesense core for the project’s long term sustainability.

Here's more background on why GPL, as described by Discourse, if you're interested: https://meta.discourse.org/t/why-gnu-license/2531. Many of the points mentioned there resonate with us.

Now, all of the above only apply to Typesense Server. Our client libraries are indeed meant to be integrated into our users’ code and so they use Apache license.

So in summary, AGPL is what is usually problematic for server software and we’ve opted not to use it. We believe GPL for Typesense Server captures the essence of what we want for this open source project. GPL has a long history of successfully being used by popular open source projects. Our libraries are still Apache licensed.

If you have specifics that prevent you from using Typesense due to a licensing issue, I'm happy to dig into this further.

**Comment by stephanoparaskeva** [comment]
Very comprehensive. It seems valid to use GPL here. I'll still choose to opt away from TypeSense as there MIT licensed projects in the same space. Awesome work regardless, and I'll  mark as a closed discussion.

**Comment by asdofindia** [comment]
> Now, if someone makes modifications to Typesense server and distributes the modifications, that's where GPL kicks in. Given that we’ve published our work to the community, we’d like for others’ modifications to also be made open to the community in the spirit of open source. We use GPL for this purpose. Other licenses would allow our open source work to be modified, made closed source and distributed, which we want to avoid with Typesense core for the project’s long term sustainability.


Would you be alright with someone modifying the Typesense server and running a hosted service over that without releasing the modified source code?

Because that's what not using AGPL means here.

**Comment by jasonbosco** [comment]
IANAL, but of course we would not be thrilled about it especially since we have a hosted version ourselves [Typesense Cloud](https://cloud.Typesense.org), but yes the license does not necessarily make that impossible to do.

**Comment by asdofindia** [comment]
Okay. Thanks for that clarification. I just didn't want typesense cloud maintainers to realize too late what mongo, elastic, etc learned about how Amazon wrecks their business model.

**Comment by stephanoparaskeva** [comment]
> Okay. Thanks for that clarification. I just didn't want typesense cloud maintainers to realize too late what mongo, elastic, etc learned about how Amazon wrecks their business model.

Mongo has been on AGPL for a long time. It didn't protect their model it only made things worse for the open source community. If amazon want to do this, nothing is stopping them. Amazon replicating this is a lesser worry than the community not wanting to use a software for fear of its license. AGPL is not something I'd want to suggest for a technology to use at all.

**Comment by jayaddison** [comment]
I'm considering switching from Elasticsearch to `typesense` by default, because while I think that Elasticsearch is going to continue to develop well in the near-term, I have a feeling that their licensing switch means that their product and company culture _may_ be heading in a more insular and defensive direction, which could indicate a shortening of their product and company's lifetime.  That's a guess; I don't know whether it'll turn out to be true.

Initially, finding that `typesense` is GPL-licensed was a plus, because I read that as a good-faith attempt to avoid the selfish-competitor concern, while also maintaining community contributions and allowing to use `typesense` as they see fit.  @kishorenc's quote in particular is reassuring:

> Given that we’ve published our work to the community, we’d like for others’ modifications to also be made open to the community in the spirit of open source.

However, some questions remain:

- What does 'distribution' mean in the terms of networked software - and isn't AGPL designed to protect exactly the kind of loophole where someone can modify code and then only offer it over-the-network so that they _don't_ have to contribute back in the spirit of open source?

- On an engineering level: what kind of modifications would teams want to make to their own `typesense` instances that would not appropriate to share back to the community?  Wouldn't it provide a useful signal to help guide product direction if "common workarounds" are made visible by default?  (this is me applying an argument something like [biological altruism](https://plato.stanford.edu/entries/altruism-biological/) to software projects)

- On an ethical level, shouldn't users have the ability to understand how their data is processed and handled by most services, since data breaches and mishandling of information are likely to become more common?  Arguably that doesn't need to be mandated in the software license, but it does make me wonder whether using GPL provides more "software gray area" for haphazard/dishonest data handling to occur within.

I've read the Discourse notes regarding AGPL v GPL and they don't seem to address the questions I had in mind.  Really they reinforce the notion that other people don't like AGPL, and some friends of theirs like GPL, and thus they'll remain with that.
